### PR TITLE
Persist TCDS FED payload - Version of 9_2_X

### PR DESCRIPTION
**Pull request by mommsen**
This pull request provides a proper unpacker for the TCDS payload and persists the relevant quantities from the TCDS FED in the TCDSRecord. This code resides in a new package DataFormats/TCDS. It replaces the FED1024 class which exposes direct access to the underlying data structure, and incorporates the separate BSTRecord. I also changed the code using the old accessors to the new ones, as far as I could find them. In addition, the EventAuxiliary is now filled with the current fill (store) number instead of 'edm::EventAuxiliary::invalidStoreNumber'.

This PR corresponds to PR #20471, but targets the 9_2_X release cycle.

### CODE REVIEW

**Comment by mommsen** [comment]
please test

**Comment by mommsen** [comment]
please test

**Comment by mommsen** [comment]
+1

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Dr15Jones** [comment]
+1
for DataFormats/StdDictionaries

**Comment by rekovic** [comment]
+1

**Comment by dmitrijus** [comment]
+1

**Comment by arunhep** [comment]
+1

**Comment by perrotta** [comment]
+1
- Mostly identical to #20471 in the master, the only difference being due to #19417 having been  integrated from 93X on, not in 92X
- No differences in jenkins outputs

**Comment by slava77** [comment]
On 9/20/17 6:51 AM, perrotta wrote:
> # No differences in jenkins outputs

but

> ROOTFileChecks: Some differences in event products or their sizes found

the default output content has changed.
Since this is a production release, deployment in T0 will likely need a
version bump in the processing.

**Comment by perrotta** [comment]
Updated my signature with the note (copy/pasted from #20471) that "a  
TCDSRecord is now persisted in the Event", which is what modifies the  
output content

Slava Krutelyov <notifications@github.com> ha scritto:

**Comment by perrotta** [comment]
Removed the "+1" from reco, so that this PR returns in our list of pending PRs until the backward compatibility issue is solved

**Comment by perrotta** [comment]
@mommsen : Just in case you still want this backport to 92X, the same fixes as in latest  #20471 should be applied here

**Comment by mommsen** [comment]
please test

File: EventFilter/Utilities/interface/FedRawDataInputSource.h [hunk 1]
```diff
@@ -47,18 +45,18 @@ friend struct InputChunk;
...
-  virtual bool checkNextEvent() override;
-  virtual void read(edm::EventPrincipal& eventPrincipal) override;
+  bool checkNextEvent() override;
+  void read(edm::EventPrincipal& eventPrincipal) override;
 
 private:
-  virtual void preForkReleaseResources() override;
-  virtual void postForkReacquireResources(std::shared_ptr<edm::multicore::MessageReceiverForSource>) override;
-  virtual void rewind_() override;
+  void preForkReleaseResources() override;
```
Comment (perrotta) [inline EventFilter/Utilities/interface/FedRawDataInputSource.h]: This method is not there in 94X: is it correct to have it here? Or, alternatively, to have it removed in 94X?

File: EventFilter/Utilities/interface/FedRawDataInputSource.h [hunk 2]
```diff
@@ -47,18 +45,18 @@ friend struct InputChunk;
...
+  void postForkReacquireResources(std::shared_ptr<edm::multicore::MessageReceiverForSource>) override;
```
Comment (perrotta) [inline EventFilter/Utilities/interface/FedRawDataInputSource.h]: This method is not there in 94X: is it correct to have it here? Or, alternatively, to have it removed in 94X?

Comment (fwyzard) [reply EventFilter/Utilities/interface/FedRawDataInputSource.h]: Looks like this has been removed from the framework in 9.3.x via da0c5e76debfb28e5594247f10809523f9927ad5 .

Comment (perrotta) [reply EventFilter/Utilities/interface/FedRawDataInputSource.h]: Ok, thank you Andrea

**Comment by perrotta** [comment]
+1
Same as  #20471, merged in 94X

**Comment by mommsen** [comment]
+1

**Comment by Martin-Grunewald** [comment]
+1

**Comment by arunhep** [comment]
+1

**Comment by Dr15Jones** [comment]
+1
For DataFormats/StdDictionaries

**Comment by rekovic** [comment]
+1

**Comment by dmitrijus** [comment]
+1

**Comment by Martin-Grunewald** [comment]
@mommsen 
Comment from ORP:
general question of who is the customer? can this not wait for 94x re-reco to avoid the issue of how to manage the data format changes?

**Comment by mommsen** [comment]
@Martin-Grunewald, there was an interest from TSG to get this online with 9_2_X. For me it would be okay to have it only in 9_4_X, assuming that it goes online next year.

**Comment by Martin-Grunewald** [comment]
@mommsen 
Do you know who specifically and for what purpose?
Thanks.

**Comment by fwyzard** [comment]
I guess if nobody insisted on the past month, we can live without for the
rest of the year...

.Andrea

**Pull request closed without merging**